model/quic-socket-tx-pfifo-scheduler.cc

```diff
@@ -0,0 +1,301 @@
+	outItem->m_isStream = true; // Packets sent with this method are always stream packets
+	outItem->m_isStream0 = false;
+	outItem->m_packet = Create<Packet>();
+	uint32_t outItemSize = 0;
+
+	while (m_appList.size() > 0 && outItemSize < numBytes) {
+		PriorityTxItem currentPriorityItem = m_appList.top();
+		currentItem = currentPriorityItem.item;
+		currentPacket = currentItem->m_packet;
+
+		NS_LOG_INFO(currentItem << " "<<GetPointer(currentPacket));
+
+		NS_LOG_INFO("Considering packet on stream "<< currentPriorityItem.streamId << " offset "<< currentPriorityItem.offset << " size "<< currentPacket->GetSize());
+
+		if (outItemSize + currentItem->m_packet->GetSize() /*- subheaderSize*/
+		<= numBytes)       // Merge
+				{
+			NS_LOG_LOGIC(
+					"Add complete frame to the outItem - size " << currentItem->m_packet->GetSize () << " m_appSize " << m_appSize);
+			QuicSubheader qsb;
+			currentPacket->PeekHeader(qsb);
+			NS_LOG_INFO(
+					"Packet: stream " <<qsb.GetStreamId() << ", offset "<<qsb.GetOffset());
+			QuicSocketTxItem::MergeItems(*outItem, *currentItem);
+			outItemSize += currentItem->m_packet->GetSize();
+
+			m_appList.pop();
+			m_appSize -= currentItem->m_packet->GetSize();
+
+			NS_LOG_LOGIC("Updating application buffer size: " << m_appSize);
+			continue;
+		} else if (firstSegment) // we cannot transmit a full packet, so let's split it and update the subheaders
+		{
+			firstSegment = false;
+
+			// subtract the whole packet from m_appSize, then add the remaining fragment (need to account for headers)
+			uint32_t removed = currentItem->m_packet->GetSize();
+			m_appSize -= removed;
+
+			// get the currentPacket subheader
+			QuicSubheader qsb;
+			currentPacket->PeekHeader(qsb);
+
+			// new packet size
+			int newPacketSizeInt = (int) numBytes - outItemSize
+					- qsb.GetSerializedSize();
+			if (newPacketSizeInt <= 0) {
+				NS_LOG_INFO("Not enough bytes even for the header");
+				m_appSize += removed;
+				break;
+			}
+			NS_LOG_INFO(
+					"Split packet on stream " <<qsb.GetStreamId() << ", sending " << newPacketSizeInt << " bytes from offset "<<qsb.GetOffset());
+
+			currentPacket->RemoveHeader(qsb);
+			uint32_t newPacketSize = (uint32_t) newPacketSizeInt;
+
+			NS_LOG_LOGIC("Add incomplete frame to the outItem");
+			uint32_t totPacketSize = currentItem->m_packet->GetSize();
+			NS_LOG_LOGIC("Extracted " << outItemSize << " bytes");
+
+			uint32_t oldOffset = qsb.GetOffset();
+			uint32_t newOffset = oldOffset + newPacketSize;
+			bool oldOffBit = !(oldOffset == 0);
+			bool newOffBit = true;
+			uint32_t oldLength = qsb.GetLength();
+			uint32_t newLength = 0;
+			bool newLengthBit = true;
+			newLength = totPacketSize - newPacketSize;
+			if (oldLength == 0) {
+				newLengthBit = false;
+			}
+			bool lengthBit = true;
+			bool oldFinBit = qsb.IsStreamFin();
+			bool newFinBit = false;
+
+			QuicSubheader newQsbToTx = QuicSubheader::CreateStreamSubHeader(
```

Comment:
This fragmentation part maybe could be encapsulated in a function for the parent